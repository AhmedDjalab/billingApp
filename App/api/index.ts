import { create } from 'apisauce';

import authStorage from '../auth/storage';
import settings from '../config/settings';

const apiService = create({
    // for android virtual device to be able to see the api - ip address
    baseURL: settings.apiUrl
});

apiService.addAsyncRequestTransform(async request => {
    const token = await authStorage.getToken();
    if (!token) return;

    request.headers['x-auth-token'] = token;
});

const { get } = apiService;



export default apiService;
