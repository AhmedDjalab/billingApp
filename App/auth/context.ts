import { createContext, useContext } from 'react';
import jwtDecode, { JwtPayload } from 'jwt-decode';


interface User {
    email: string;
    iat: number;
    name: string;
    userId: number;
}

interface ContextType {
    user: User;
    setUser: (user: Object | null) => void;
    logout: () => void;
}

const AuthContext = createContext<ContextType | any>({ user: {} });

const useAuthContext = () => {
    const { user, setUser } = useContext<ContextType>(AuthContext);

    const login = (token: string) => {
        const currentUser = jwtDecode<JwtPayload>(token);
        setUser(currentUser);
    };

    const logout = () => {
        setUser(null);
    };

    return { user, login, logout };
};

export { AuthContext, useAuthContext };
