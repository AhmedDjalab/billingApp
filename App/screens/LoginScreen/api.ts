import apiService from '../../api/index';

const login = (email: string, password: string) =>
{
return apiService.post('/users/authenticate',
 { "email":email, "password":password });}


export default { login };
