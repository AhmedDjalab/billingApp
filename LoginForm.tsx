import React from "react";
import { View } from "react-native";
import { ErrorMessage, Form, FormField, SubmitButton } from "./src/components/Forms";

export default function LoginForm({ handleSubmit, validationSchema, loginFailed }) {
  return (<Form 
    
    initialValues={{
    email: '',
    password: ''
  }} onSubmit={handleSubmit} validationSchema={validationSchema}>
    <ErrorMessage error="Invalid email and/or password" visible={loginFailed} />
    <View style={{
      paddingBottom: 20
    }}>
      <FormField testID="email" autoCapitalize="none" autoCorrect={false} icon="email" keyboardType="email-address" name="email" placeholder="Email" textContentType="emailAddress" />
    </View>
    <View style={{
      paddingBottom: 20
    }}>
      <FormField testID="password" autoCapitalize="none" autoCorrect={false} icon="lock" name="password" placeholder="Password" secureTextEntry textContentType="password" />
    </View>
    <SubmitButton testID="LoginButton" title="Login" contained={true} />
  </Form>);
}

