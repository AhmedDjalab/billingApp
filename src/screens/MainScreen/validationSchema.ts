import * as Yup from 'yup';

export default Yup.object().shape({
    label: Yup.string().required().min(1).label('Product Label'),
    price: Yup.number().required().min(1).label('Price'),
    quantity: Yup.number().required().min(1).label('Quantity'),

});
