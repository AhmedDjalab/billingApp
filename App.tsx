import { NavigationContainer } from '@react-navigation/native';
import React from 'react';
import { useState } from 'react';
import { Platform, StyleSheet, Text, View } from 'react-native';
import { AuthContext, useAuthContext } from './src/auth/context';
import LoginScreen from './src/screens/LoginScreen';
import MainScreen from './src/screens/MainScreen/mainScreen';
import { createStackNavigator } from '@react-navigation/stack';
import authStorage from './src/auth/storage';
import AppLoading from 'expo-app-loading';
import navigationTheme from './src/navigation/navigationTheme';
import { MaterialCommunityIcons } from '@expo/vector-icons';

export default function App() {
  const [user, setUser] = useState(null);
  const { logout } = useAuthContext();

  const [isReady, setIsReady] = useState(false);
  const Stack = createStackNavigator() ;
  const restoreUser = async () => {
    const currentUser = await authStorage.getUser();
    console.log("this is from cach " , currentUser)
    if (currentUser) setUser(currentUser);
};

const handleLoadingError = () => {
    throw new Error('App loading error');
};
const handleLoadingFinish = () => setIsReady(true);
if (!isReady)
        return (
            <AppLoading
                startAsync={restoreUser}
                onError={handleLoadingError}
                onFinish={handleLoadingFinish}
            />
        );
  return (
   
    <View style={styles.container}>
    <AuthContext.Provider value={{ user, setUser }}>
       <NavigationContainer  theme={navigationTheme}> 
       {user ? <Stack.Navigator >
         <Stack.Screen  
       component={MainScreen} 
       name={"Billing"}
       options={{
        title: "Billing",
        headerTitleAlign: 'center',
        headerRight : () =>(
          <View style={styles.iconContainer}>
           <MaterialCommunityIcons
                            color="red"
                            name="logout"
                            size={30}
                            onPress={()=> setUser(null)}
                            
                        />
          
          </View>
        )
      }}
       
       /></Stack.Navigator> :<LoginScreen />}
       </NavigationContainer>
    </AuthContext.Provider>
</View>
  
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
   
    backgroundColor: '#fff',
   
  },
  iconContainer: {
    flexDirection: "row",
    justifyContent: "space-evenly",
    width: 120
  }
});
