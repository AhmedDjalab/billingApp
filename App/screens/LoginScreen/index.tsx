import React, { useState } from 'react';
import { Image  , SafeAreaView, View} from 'react-native';
import { ApiResponse } from 'apisauce';

import {
    Form,
    ErrorMessage,
    FormField,
    SubmitButton
} from '../../components/Forms';
import validationSchema from './validationSchema';
import auth from './api';
import { useAuthContext } from '../../auth/context';
import UserCredentials from './model';
import styles from './styles';

export default function LoginScreen() {
    const { login } = useAuthContext();
    const [loginFailed, setLoginFailed] = useState(false);

    const handleSubmit = async ({ email, password }: UserCredentials) => {
      try { const result: ApiResponse<any> = await auth.login(email, password);
         
        if (!result.ok) {
            setLoginFailed(true);
            return;
        }
        setLoginFailed(false);
        console.log("thos os data login " ,result.data["token"] )
        login( result.data["token"] );
    
    
    } catch (e){
            console.log("this is errros " , e)
        }
    };

    return (
        <SafeAreaView style={{paddingTop:40 , flex:1}}>

       
       
       <View style={styles.container}>
            {/* <Image
                source={require('../assets/logo-red.png')}
                style={styles.logo}
            /> */}
            <Form
                initialValues={{ email: '', password: '' }}
                onSubmit={handleSubmit}
                validationSchema={validationSchema}>
                <ErrorMessage
                    error="Invalid email and/or password"
                    visible={loginFailed}
                />
              <View style={{paddingBottom:20}}>
                   <FormField
                  
                    autoCapitalize="none"
                    autoCorrect={false}
                    icon="email"
                    keyboardType="email-address"
                    name="email"
                    placeholder="Email"
                    textContentType="emailAddress"
                />
              </View>
              <View style={{paddingBottom:20}}>
                <FormField
                    autoCapitalize="none"
                    autoCorrect={false}
                    icon="lock"
                    name="password"
                    placeholder="password"
                    secureTextEntry
                    textContentType="password"
                />
                </View>
                <SubmitButton title="Login" contained={true}/>
            </Form>
        </View>
        </SafeAreaView>
    );
}
