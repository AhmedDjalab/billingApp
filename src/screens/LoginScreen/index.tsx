import React, { useState } from 'react';
import { Image  , SafeAreaView, View} from 'react-native';
import { ApiResponse } from 'apisauce';

import {
    Form,
    ErrorMessage,
    FormField,
    SubmitButton
} from '../../components/Forms';
import validationSchema from './validationSchema';
import auth from './api';
import { useAuthContext } from '../../auth/context';
import UserCredentials from './model';
import styles from './styles';
import LoginForm from '../../../LoginForm';

export default function LoginScreen() {
    const { login } = useAuthContext();
    const [loginFailed, setLoginFailed] = useState(false);

    const handleSubmit = async ({ email, password }: UserCredentials) => {
      try { 
          
        const result: ApiResponse<any> = await auth.login(email, password);
         console.log("ths is daaa loging g " , email , password)
        if (!result.ok) {
            setLoginFailed(true);
            return;
        }
        setLoginFailed(false);
        console.log("thos os data login " ,result.data["token"] )
        login( result.data["token"] );
    
    
    } catch (e){
            console.log("this is errros " , e)
        }
    };

    return (
        <SafeAreaView style={{paddingTop:40 , flex:1}}>

       
       
       <View style={styles.container}>
          <LoginForm   handleSubmit={handleSubmit} validationSchema={validationSchema} loginFailed={loginFailed}  />
        </View>
        </SafeAreaView>
    );
}
