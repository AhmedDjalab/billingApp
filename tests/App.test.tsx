
import React from 'react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import App from '../App';
import Enzyme , {mount, shallow} from 'enzyme';
import { render, fireEvent, waitFor , act } from '@testing-library/react-native';
import user from '@testing-library/user-event'


import Adapter from 'enzyme-adapter-react-16';
import LoginScreen from '../src/screens/LoginScreen';
import { AuthContext , useAuthContext } from '../src/auth/context';
import LoginForm from '../LoginForm';
import validationSchema from '../src/screens/MainScreen/validationSchema';
import { Formik } from 'formik';

Enzyme.configure({ adapter: new Adapter() });
let realUseContext;
let useContextMock;
let loggedUser = null ; 
// Setup mock
beforeEach(() => {
    realUseContext = React.useContext;
    useContextMock = React.useContext = jest.fn();
});
// Cleanup mock
afterEach(() => {
    React.useContext = realUseContext;
});


describe('<App />', () => {
  it('renders correctly, test using Jest + Enzyme', () => {
    expect(mount(
      <AuthContext.Provider value={{ user: loggedUser ,  useContextMock }}>
    <App/>
    </AuthContext.Provider>
    )).toMatchSnapshot();
  });

});
it('does not show error messages when input values are valid', async () => {

  const contextValues = { user: null };
  let promise = Promise.resolve() ; 
  const handleSubmit = jest.fn(()=> Promise.resolve())
  jest.spyOn(React, 'useContext').mockImplementation(() => ({
    user: null
  }));
  const myInitialState = false;

  React.useState = jest.fn().mockReturnValue([myInitialState, {}])

  const wrapper = shallow(<LoginForm
     handleSubmit={handleSubmit}
     validationSchema={validationSchema}
     loginFailed={false}
  />);
  let email = wrapper.find(`[testID="email"]`)
  let pass = wrapper.find(`[testID="password"]`)

  let submit = wrapper.find(`[type="submit"]`)
 await act(()=> promise )  ; 

 
  
});
it('renders correctly, test using Jest + Enzyme', () => {
  expect(mount(
    <AuthContext.Provider value={{ user: loggedUser ,  useContextMock }}>
  <App/>
  </AuthContext.Provider>
  )).toMatchSnapshot();
});


// it('is calling handle sumbit when click on button', async () => {

// const contextValues = { user: null };
// let promise = Promise.resolve() ; 
// let handleSubmit = jest.fn();
// jest.spyOn(React, 'useContext').mockImplementation(() => ({
//   user: null
// }));
// const myInitialState = false;

// React.useState = jest.fn().mockReturnValue([myInitialState, {}])
//   LoginForm.prototype.handleSubmit = jest.fn();

// const wrapper = shallow(<LoginForm
//    handleSubmit={handleSubmit}
//    validationSchema={validationSchema}
//    loginFailed={false}
// />);

// let email = wrapper.find(`[testID="email"]`)
// let pass = wrapper.find(`[testID="password"]`)
// let submit = wrapper.find(`[testID="LoginButton"]`);
// email.simulate('change', {
//   target: {
//       value: 'miqueias@gmail.com'
//   }
// });
// pass.simulate('change', {
// target: {
//     value: '122hghghgh3'
// }
// });
// console.log("thid is butons " , {...submit})
// await waitFor(() => {
//   submit.simulate("submit");});



 

// expect(handleSubmit).toBeCalled();
 



// });