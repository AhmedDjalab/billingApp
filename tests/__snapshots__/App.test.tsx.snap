// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly, test using Jest + Enzyme 1`] = `ReactWrapper {}`;

exports[`Form Table  renders correctly, test using Jest + Enzyme 1`] = `ReactWrapper {}`;

exports[`renders correctly, test using Jest + Enzyme 1`] = `ReactWrapper {}`;
